src/core/project/qgsproject.cpp

```diff
@@ -1305,6 +1308,87 @@ void QgsProject::setAvoidIntersectionsMode( const Qgis::AvoidIntersectionsMode m
   emit avoidIntersectionsModeChanged();
 }
 
+static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFlags projectReadFlags, Qgis::ProjectFlags projectFlags )
+{
+  QgsMapLayer::ReadFlags layerFlags = QgsMapLayer::ReadFlags();
+  if ( projectReadFlags & Qgis::ProjectReadFlag::DontResolveLayers )
+    layerFlags |= QgsMapLayer::FlagDontResolveLayers;
+  // Propagate trust layer metadata flag
+  if ( ( projectFlags & Qgis::ProjectFlag::TrustStoredLayerStatistics ) || ( projectReadFlags & Qgis::ProjectReadFlag::TrustLayerMetadata ) )
+    layerFlags |= QgsMapLayer::FlagTrustLayerMetadata;
+  // Propagate open layers in read-only mode
+  if ( ( projectReadFlags & Qgis::ProjectReadFlag::ForceReadOnlyLayers ) )
+    layerFlags |= QgsMapLayer::FlagForceReadOnly;
+
+  return layerFlags;
+}
+
+void QgsProject::preloadProviders( const QVector<QDomNode> &asynchronusLayerNodes,
+                                   const QgsReadWriteContext &context,
+                                   QMap<QString, QgsDataProvider *> &loadedProviders,
+                                   QgsMapLayer::ReadFlags layerReadFlags,
+                                   int totalProviderCount )
+{
+
+  QVector<QgsRunnableProviderCreator *> runnables;
+  QThreadPool threadPool;
+
+  QgsSettings settings;
+  if ( settings.contains( QStringLiteral( "provider-preloading/max-thread-count" ) ) )
+  {
+    threadPool.setMaxThreadCount( settings.value( QStringLiteral( "provider-preloading/max-thread-count" ) ).toInt() );
+  }
+  else
+  {
+    threadPool.setMaxThreadCount( QThread::idealThreadCount() );
+    settings.setValue( QStringLiteral( "provider-preloading/max-thread-count" ), threadPool.maxThreadCount() );
```

Comment:
Do we need to set the value here to settings? I would prefer to keep it unset...

Reply:
But where and when  we can set it?

Reply:
So... are we setting it here just so that it would appear in options dialog in the advanced tab? if so, then I would suggest to move setting this value to QgsOptions... (or how is it done with other advanced config options?)